### ISSUE TITLE
no-magic-numbers triggering for array access keys

### ISSUE THREAD

**Issue by cleavera** [palantir/tslint#2721]
### Bug Report

- __TSLint version__: 5.2.1
- __TypeScript version__: 2.2.0
- __Running TSLint via__: gulp-tslint

#### TypeScript code being linted

```ts
const match: RegExpMatchArray = require.match(/^(\^\^?)?(\?)?(\^\^?)?/);
const inheritType: string = match[3];
const isOptional: boolean = !!match[2];
```

with `tslint.json` configuration:

```json
"no-magic-numbers": true,
```

#### Actual behavior
```'magic numbers' are not allowed``` fired for both array access keys
match[__3__];
match[__2__];

#### Expected behavior
Array access keys are not magic numbers so should not be flagged as such when linting

**Comment by ajafff** [comment]
what makes a number used as array index different from a number used elsewhere? they are both magic numbers.

**Comment by cleavera** [comment]
Not really, I wouldn't call a property key on an object a magic string, so what make arrays special?

What would you propose?
```
const match: RegExpMatchArray = require.match(/^(\^\^?)?(\?)?(\^\^?)?/);
const isOptionalIndexKey: number = 2;
const isOptional: boolean = !!match[isOptionalIndexKey];
```

It doesn't provide anything extra (in readability or maintainability), and make the rule a bit of a pig to use.

**Comment by adidahiya** [comment]
How often does this show up in your codebase @cleavera? Most of the examples I've seen so far have been contrived, except for the regexp match, which seems like something you can special case with a `// tslint:disable` flag.

**Comment by cleavera** [comment]
By far and away the biggest usages are in unit tests, where the code pushes different items to an array in different circumstances and the unit tests are checking that specific items in the array equal specific values.
```ts
it('should provide a link to the landing page', function() {
  expect($ctrl.actions[0]).toEqual({ state: 'core-landingPage' });
});

it('should provide a link to the person summary', function() {
  expect($ctrl.actions[1]).toEqual({ state: 'hrm-person', params: { personId: '123' } });
});

it('should provide a link to create a new person', function() {
  expect($ctrl.actions[2]).toEqual({ state: 'hrm-person-details', params: { personId: 'new' } });
});
```

In this case the numbers literally represent the number. I could have variables for the them but they provide no value whatsoever.

I would be interested to see a case where extracting the number to a named variable provides some extra level of maintainability or readability.

If I take ajafff's (rather abstract) example from the pull request and change it to conform to the rule:
```ts
const five: number = 5,
  ten: number = 10,
  two: number = 2;

doStuff(arr[five], arr[ten], arr[two]);
```

When accessing items from the array the number isn't magical, it literally represents the number, when you access a specific item from an array you're saying I want the 5th item in the array, you're not saying i want the [abstractConcept] item in the array.

If I take us back to the real use cases for this rule:
```ts
const total: number = 60 * 60 * 60 * 24;
```

You can see how extracting those numbers provides value to readability, I just don't believe this is the case with accessing items from an array

**Comment by roman-petrov** [comment]
agree with @ajafff, array indices are the same numbers as other and I think is not a good practice to allow code with somewhat random numbers appearing in array access expressions. For example, for statically indexed arrays, when we know array size at run time: it is often better to use classes and named properties instead this array than use array indices. For example:
```
const person: string[2] = [];
person[0] = 'James';
person[1] = 'Hetfield';
person[2] = '<address>';

// Somewhere below or even in another function
const address: string = person[2];
```
So if we allow magic numbers as array indices, code above will be considered as valid, but actually it is a bad code.

**Comment by cleavera** [comment]
I don't know how banning magic number for array access makes that code any better?

```
const person: string[2] = [];
person[0] = 'James';
person[1] = 'Hetfield';
person[2] = '<address>';

const addressIndex: number = 2;

// Somewhere below or even in another function
const address: string = person[addressIndex];
```

It's still not great code and it has no more meaning than the example you posted

**Comment by cleavera** [comment]
If i can use the comparison with objects:
```ts
let obj = {
  'my-key': 'Some value'
};

console.log(obj['my-key']);
```
Do you consider 'my-key' to be a magic string? Of course not its just the name of a property on the object. Thats all the number is for array value access, the name of a property on the array.

**Comment by roman-petrov** [comment]
@cleavera, actually my answer to your last question would be 'yes'. In my projects I prefer do not use object access by statically typed string keys because it is not type-safe and hard to change when you start refactor.

**Comment by cleavera** [comment]
I do agree to an extent, I wouldnt voluntarily use the square bracket notation, however in the cases where you do need to there is nothing to be gained from extracting that string into a variable.

**Comment by cleavera** [comment]
Are we at least open to the idea of having this as a configuration option similar to how eslint works (http://eslint.org/docs/rules/no-magic-numbers)?

**Comment by roman-petrov** [comment]
My opinion is than configuration option(s) (not enabled if not specified explicitly) would be much better, than making rule 'soft' by default.

**Comment by cleavera** [comment]
I still think the majority of people will enable this option by default however I'm happy with this compromise. Am I okay to update my PR and resubmit?

**Comment by adidahiya** [comment]
@cleavera yeah that sounds good as a new rule option.

**Comment by talmukund** [comment]
Can I generate new PR since old one was rejected. I will implement same logic.

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!
